determined-ai/determined
# feat: backend support for inference metric tracking part 2

## Description

[MLG-722]

Adds support to the linking and reporting metrics to specific checkpoints or model versions. This is to allow users to utilize the generic metrics framework to assist in writing inference experiments.

## How to use:

### Inside Inference Experiment

Reference the checkpoint or model version and link to this inference run:

```python3
import determined as det
from determined.experimental import client

with det.core.init() as core_context:

    # Checkpoint
    ckpt = client.get_checkpoint("151edd59-771e-4d97-b660-620316c1993f")
    core_context.experimental.report_task_using_checkpoint(ckpt)
    
    # Or Model Version
    model = client.get_model("cifar_inference")
    model_version = model.get_version(1)
    core_context.experimental.report_task_using_model_version(model_version)
```

During the run, you may report custom inference metrics like so:
```python3
core_context.train.report_metrics(
    steps_completed=0, 
    group="inference",  # This group string can be anything. Use it again when looking up
    metrics={
        "inf_val_1": value1,
        "inf_val_2": value2,
    },
)
```

### After a run

You may query for all inference metrics linked to a particular checkpoint or model version by querying the object in question with the SDK and calling `<OBJECT>.get_metrics()`

```python3
from determined.experimental import client

# Checkpoint
ckpt = client.get_checkpoint("151edd59-771e-4d97-b660-620316c1993f")
metrics = ckpt.get_metrics("inference")

# Or Model Version
model = client.get_model("cifar_inference")
model_version = model.get_version(1)
metrics = model_version.get_metrics("inference")
```

The Metrics come back as an `TrialMetrics` Iterable object in the same style as our training and validation metrics. You may use any group name in the `get_metrics` call with "" meaning "all metrics" (including training and validation).



## Test Plan

1. Pull latest `main` for the updated examples. Run `determined/examples/computer_vision/cifar10_pytorch` with `det e create const.yaml`
2. Create a `Model` in the cluster. You can do so from `<DET_MASTER>/det/models` and clicking "New Model"
3. Navigate to the experiment from (1), and save some checkpoints to the `Model` in (2)
4. Back in `determined/examples/computer_vision/cifar10_pytorch`, edit `distributed_inference_example.yaml` to point to the name of the `Model` in (2) and change the model versions to be valid to however many you created.
5. Run `det e create distributed_inference_example.yaml .`
6. Start a Jupyter Notebook on Determined
7. Run 
```python3
from determined.experimental import client

model = client.get_model("<MY_MODEL_VERSION_NAME>")
model_version = model.get_version(1)

for metric in model_version.get_metrics():
    print(metric)
```
8. You should get back some `TrialMetric` objects showing the inference metrics associated with the `ModelVersion`






## Commentary (optional)





## Checklist

  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.

## Ticket






[MLG-722]: https://hpe-aiatscale.atlassian.net/browse/MLG-722?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
